docs/proposals/kubelet-systemd.md

```diff
@@ -0,0 +1,277 @@
+
+`systemd` defines a set of slices.  By default, service and scope units are placed in
+`system.slice`, virtual machines and containers registered with `systemd-machined` are
+found in `machine.slice`, and user sessions handled by `systemd-logind` in `user.slice`.
+
+## kubelet cgroup driver
+
+The current implementation of the `kubelet` reads and writes to the `cgroup` file-system
+directly during initial bootstrapping in ways that are not appropriate for `systemd`
+systems.
+
+This proposal proposes that the `kubelet` accept a new flag:
+
+* `--cgroup-driver="":` - cgroup driver used by the kubelet.
+Possible values: `cgroupfs` or `systemd`.
+
+The `ContainerManager` interface should support an implementation per driver option.
+
+Shared code across implementations should be split into a common location.
+
+By default, the `kubelet` should default `--cgroup-driver` to `systemd` on `systemd` distributions.
+
+## kubelet cgroup bootstrapping under systemd
+
+The following sections outline how a `systemd` system supports local node accounting
+requirements.
+
+### Node capacity
+
+The `kubelet` will continue to interface with `cAdvisor` to determine node capacity.
+
+### System reserved
+
+The node may set aside a set of designated resources for non-Kubernetes components.
+
+The `kubelet` accepts a flag of the following form to reserve compute resources
+for non-Kubernetes components:
+
+* `--system-reserved=:` - A set of `ResourceName`=`ResourceQuantity` pairs that
+describe resources reserved for host daemons.
+
+The `kubelet` does not enforce `--system-reserved`, but the ability to distinguish
+the static reservation from observed usage is important for node accounting.
+
+The `kubelet` will accept a new flag in a `systemd` environment to support accounting
+system-reserved resources.
+
+* `--slice-system="system.slice"` - Slice reserved for
+system usage when using systemd cgroup driver.
+
+The `kubelet` will error in any of the following conditions:
+* the named `slice` does not exist
+* if the `cpu` and `memory` controllers are not enabled on the specified slice
+
+### Kubernetes reserved
+
+The node may set aside a set of resources for Kubernetes components: 
+
+* `--kube-reserved=:` - A set of `ResourceName`=`ResourceQuantity` pairs that
+describe resources reserved for host daemons.
+
+The `kubelet` does not enforce `--kube-reserved` at this time, but the ability
+to distinguish the static reservation from observed usage is important for node accounting.
+
+This proposal asserts that all components that Kubernetes brings to the node are
+parented by a common `slice` in a separate part of the hierarchy than system-reserved
+resources.
+
+This proposal asserts that `kubernetes.slice` is the default slice associated with
+the `kubelet` and `kube-proxy` service units defined in the project.
+
+The `kubelet` will detect the parent `slice` from which it was launched to track
```

Comment:
> This can be achieved through cgroupfs directly instead of having to understand systemd APIs.

_can_, but this doesn't imply _should_ ;-). More reasoning/justification would be nice. Is your main concern maintaining a separate codepath for non-systemd systems?